ros/src/waypoint_updater/waypnt_updater.py

```diff
@@ -837,126 +810,100 @@ def set_waypoints_velocity(self):
                                  lookahead_wps)
-                else:
-                    rospy.logdebug("Within buffer, but not travelling at creeping speed")
-                    
-                    recalc = self.produce_slowdown(self.final_waypoints_start_ptr,
-                        self.lookahead_wps,
-                        dist_to_tl - (self.dyn_tl_buffer - 1.0))
-
-            elif self.state == 'slowdown' and self.stop_target == self.next_tl_wp:
-                    recalc = self.produce_slowdown(self.final_waypoints_start_ptr,
-                                               self.lookahead_wps,
-                                               dist_to_tl - (self.dyn_tl_buffer - 1.0))
-
-            elif dist_to_tl < self.dyn_creep_zone + self.dyn_tl_buffer:
-                if self.waypoints[self.final_waypoints_start_ptr].get_v() <=\
-                        self.handoff_velocity:
-                    self.set_creep(self.final_waypoints_start_ptr, self.lookahead_wps)
-                else:
-                    if self.state != 'slowdown' and self.state != 'stopped':
-                        rospy.logwarn("now start to slowdown")
-                    recalc = self.produce_slowdown(self.final_waypoints_start_ptr,
-                                               self.lookahead_wps,
-                                               dist_to_tl - (self.dyn_tl_buffer - 1.0))
-                    self.stop_target = self.next_tl_wp
             else:
-                # not sure when this might trigger but not match next if statement - then 
-                # reduce car velocity to 0.0 stoping before lights
-                # if self.state != 'slowdown' and self.state != 'stopped':
-                #    rospy.logwarn("set car state to slowdown at ptr = {}".format(self.final_waypoints_start_ptr))
-
-                if dist_to_tl - self.dyn_tl_buffer > self.min_stop_distance:
-                    if self.state != 'slowdown':
-                        rospy.logwarn("Within stopping distance of TL - Start to slowdown!")
-                        self.stop_target = self.next_tl_wp
-                    else:
-                        rospy.logwarn("we should not get here! {}, {}".format(self.stop_target, self.next_tl_wp))
-                    recalc = self.produce_slowdown(self.final_waypoints_start_ptr,
-                                               self.lookahead_wps,
-                                               dist_to_tl - (self.dyn_tl_buffer - 1.0))
+                recalc = self.produce_slowdown(self.final_waypoints_start_ptr,
+                                           self.lookahead_wps,
+                                           dist_to_tl - (self.dyn_tl_buffer - self.dyn_buffer_offset))
+
+        # switch to slowdown if possible!
+        elif (self.state == 'speedup' or self.state == 'maintainspeed') and\
+               dist_to_tl - self.dyn_tl_buffer < self.stopping_distance :
```

Comment:
The stopping transition seems fixed with the updated buffer parameter at 2.5m.

However, during some various testing I got another fell out of state machine error for the case of moving the target waypoint while the car was already in slowdown where the new target distance was shorter than stopping_distance, but it couldn't readjust the slowdown because the condition here requires speedup or maintainspeed state.

My test pattern may not be actually realistic, but maybe we need to add a catch for the possibility of being in 'slowdown' state condition here?

Reply:
Added slowdown to this check.